File: fix.dart
import 'dart:io';

void main() async {
  final outTxt = File('out_utf8.txt');
  if (!await outTxt.exists()) {
    print("out_utf8.txt not found");
    return;
  }

  final filesWithDups = <String>{};
  final lines = await outTxt.readAsLines();
  for (final line in lines) {
    if (line.contains('duplicate_constructor')) {
      final parts = line.split(' - ');
      if (parts.length >= 2) {
        final filePath = parts[2].split(':')[0].trim();
        filesWithDups.add(filePath);
      }
    }
  }

  for (final filePath in filesWithDups) {
    final file = File(filePath);
    if (!await file.exists()) continue;

    final fileLines = await file.readAsLines();
    final classNames = <String>[];

    for (final l in fileLines) {
      final m = RegExp(r'^\s*class\s+(\w+)\s+extends').firstMatch(l);
      if (m != null) {
        classNames.add(m.group(1)!);
      }
    }

    bool changed = false;
    for (final cls in classNames) {
      int idxConstEmpty = -1;
      int idxNonConstEmpty = -1;
      int idxOther = -1;

      for (int i = 0; i < fileLines.length; i++) {
        final stripped = fileLines[i].trim();
        if (stripped == "const $cls({super.key});") {
          idxConstEmpty = i;
        } else if (stripped == "$cls({super.key});") {
          idxNonConstEmpty = i;
        } else if (stripped.startsWith("$cls(") ||
            stripped.startsWith("const $cls(")) {
          if (stripped != "const $cls({super.key});" &&
              stripped != "$cls({super.key});") {
            idxOther = i;
          }
        }
      }

      if (idxConstEmpty != -1 && idxNonConstEmpty != -1) {
        fileLines[idxNonConstEmpty] = '///DELETE_ME///';
        changed = true;
      } else if (idxConstEmpty != -1 && idxOther != -1) {
        fileLines[idxConstEmpty] = '///DELETE_ME///';
        changed = true;
      } else if (idxNonConstEmpty != -1 && idxOther != -1) {
        fileLines[idxNonConstEmpty] = '///DELETE_ME///';
        changed = true;
      }
    }

    if (changed) {
      final newLines = fileLines.where((l) => l != '///DELETE_ME///').toList();
      await file.writeAsString(newLines.join('\n') + '\n');
      print("Fixed $filePath");
    }
  }
}

File: fix.py
import os
import re

def fix():
    files_with_dups = set()
    with open('out_utf8.txt', 'r', encoding='utf8') as f:
        for line in f:
            if 'duplicate_constructor' in line:
                parts = line.split(' - ')
                if len(parts) >= 2:
                    file_path = parts[1].split(':')[0].strip()
                    files_with_dups.add(file_path)
    
    for file_path in files_with_dups:
        if not os.path.exists(file_path):
            continue
        with open(file_path, 'r', encoding='utf8') as f:
            lines = f.readlines()
        
        # Find all classes
        class_names = []
        for line in lines:
            m = re.match(r'^\s*class\s+(\w+)\s+extends', line)
            if m:
                class_names.append(m.group(1))
        
        # To decide which constructor to delete, we find them first
        for cls in class_names:
            idx_const_empty = -1
            idx_non_const_empty = -1
            idx_other = -1
            
            for i, line in enumerate(lines):
                stripped = line.strip()
                if stripped == f"const {cls}({{super.key}});":
                    idx_const_empty = i
                elif stripped == f"{cls}({{super.key}});":
                    idx_non_const_empty = i
                elif stripped.startswith(f"{cls}(") or stripped.startswith(f"const {cls}("):
                    if stripped != f"const {cls}({{super.key}});" and stripped != f"{cls}({{super.key}});":
                        idx_other = i
                        
            # Delete logic
            if idx_const_empty != -1 and idx_non_const_empty != -1:
                # Delete non_const empty
                lines[idx_non_const_empty] = None
                print(f"Deleted non-const empty constructor in {file_path}")
            elif idx_const_empty != -1 and idx_other != -1:
                # Delete empty const because the other has params
                lines[idx_const_empty] = None
                print(f"Deleted const empty constructor in {file_path}")
            elif idx_non_const_empty != -1 and idx_other != -1:
                lines[idx_non_const_empty] = None
                print(f"Deleted non-const empty constructor in {file_path}")

        with open(file_path, 'w', encoding='utf8') as f:
            for line in lines:
                if line is not None:
                    f.write(line)

fix()
